app/controllers/concerns/demo_mode/demoable.rb

```diff
@@ -28,7 +28,7 @@ def _demo_mode_logout_check!
     end
 
     def demo_splash!
-      if _demo_mode_current_signinable.blank? && !current_demo_session&.custom_sign_in?
+      if current_demo_session.nil? || (_demo_mode_current_signinable.blank? && !current_demo_session&.custom_sign_in?)
```

Comment:
The one thing to make sure we test here would be the "sign up link" feature, which allows the splash page to link out to an unauthenticated sign up flow. [Per the Readme](https://github.com/Betterment/demo_mode?tab=readme-ov-file#the-sign-up-link), I would not expect this to break in a properly-configured downstream app, but if you already have it set up to test locally, I'd just want to confirm.

I think the "Display Credentials" feature would be fine here as well, but similar to the sign up link, it allows you to link out to the consuming app's own login screen so that you can exercise real login behavior, and I would want to make sure that this feature continues to work as well.

Reply:
@smudge I checked both features in retail. They work. However, I think it's because, retail circumvents Deomable's  action hooks when the user is already logged in (retail adds Demoable to `App::LoginsController` and this controller has other hooks which will redirect before reaching Demoable's hooks).

Nevertheless, I believe this line would have caused issues in the features you mentioned. My interpretation of the code is that `session[:demo_session]` wouldn't exist if we authenticate through "Sign up link" or with "Display Credentials". Because of that, we would have been logged out automatically because `current_demo_session` would be `nil`. 

As a result, I modified the implementation. It now clears `session[:demo_session]` and signs out the user only when a `session[:demo_session]` exists AND the session record it references is not in the database.

Reply:
Great, makes sense. I like the new, more-restrictive implementation. I see less room for unintended consequences, because it is fully conditional on the `find_by` failing to find an ID that exists in the session.